Firestore/Swift/Source/Codable/third_party/FirestoreEncoder.swift

```diff
@@ -0,0 +1,616 @@
+// This file is derived from swift/stdlib/public/SDK/Foundation/JSONEncoder.swift
+// and swift/stdlib/public/SDK/Foundation/PlistEncoder.swift
+
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import FirebaseFirestore
+
+extension CollectionReference {
+  public func addDocument<T: Encodable>(_ item: T) -> DocumentReference {
+    do {
+      return addDocument(data: try Firestore.Encoder().encode(item))
+    } catch let error {
+      fatalError("Unable to encode data with Firestore encoder: \(error)")
+    }
+  }
+
+  public func addDocument<T: Encodable>(_ item: T, _ completion: ((Error?) -> Void)?) -> DocumentReference {
+    do {
+      let encoded = try Firestore.Encoder().encode(item)
+      return addDocument(data: encoded, completion: completion)
+    } catch let error {
+      Firestore.firestore().settings.dispatchQueue.sync {
+        completion!(error)
+      }
+      return document() // Is there something better to return after the error?
```

Comment:
This particular wart exposes a major problem. As I'm messing around with this I think the inconsistent failure behavior between completion-accepting and non-completion-accepting variants of these methods is wrong. It makes it difficult to describe what our expectations are on conforming types, and makes it difficult to concisely describe the API guarantee we're making.

I'd like us to say simply that encoding errors on any of these encodable-accepting types are fatalErrors, and then we don't need to synthesize fake instances to return like this.

Users that want to handle the error programmatically can use the encode function (or encoder type) directly.

Reply:
Changed to always fatalError